tools/hdf_dump_translation_dataset.py

```diff
@@ -0,0 +1,321 @@
+    """
+    lengths = numpy.array([source_lengths, target_lengths], dtype="int32").transpose()
+
+    offset = self._number_of_processed_lines
+    assert len(lengths) + offset <= self.number_of_lines, "More lines in the corpus files than specified."
+
+    self.hdf_file["seqLengths"][offset:offset + len(lengths), :] = lengths
+
+  def _write_tags(self, chunk_size):
+    """
+    Writes the sequence tags to the HDF file.
+
+    :param int chunk_size: number of lines in the current chunk
+    """
+    offset = self._number_of_processed_lines
+
+    tags = [numpy.array("line-" + str(offset + i), dtype=self._tag_dtype, ndmin=1) for i in range(chunk_size)]
+    tags = numpy.concatenate(tags)
+
+    self.hdf_file["seqTags"][offset: offset + chunk_size] = tags
+
+  def _write_data_indices(self, chunk, side):
+    """
+    Writes the main data (word indices for the source or target corpus) to the HDF file.
+
+    :param list[numpy.ndarray] chunk: word indices for all lines in the current chunk
+    :param str side: "source" or "target"
+    """
+    indices = numpy.concatenate(chunk)
+
+    if side == "source":
+      dataset = self.hdf_file["inputs/"]
+    else:
+      dataset = self.hdf_file["targets/data/classes"]
+
+    offset = self._write_offsets[side]
+    length = len(indices)
+    if offset + length > len(dataset):
+      buffer_size = max(self.data_buffer_size, length)
+      dataset.resize((offset + buffer_size,))
+
+    dataset[offset:offset + length] = indices
+
+    self._write_offsets[side] += length
+
+  def _finalize_data(self):
+    """
+    Called after all data is written. Checks number of lines and resizes datasets down to actual data size.
+    """
+    # Make sure the number of lines given by the user was correct.
+    # Otherwise lengths and labels would have trailing zeros.
+    assert self.number_of_lines == self._number_of_processed_lines, "Fewer lines ({}) in the corpus files " \
+        "than specified ({}).".format(self._number_of_processed_lines, self.number_of_lines)
+
+    # Trim datasets to actually occupied length, i.e. remove unused reserved space.
+    self.hdf_file["inputs"].resize((self._write_offsets["source"],))
+    self.hdf_file["targets/data/classes"].resize((self._write_offsets["target"],))
+
+  def _line_to_indices(self, line, side):
+    """
+    Converts a line of text to an array of word indices.
+
+    :param str line: input line
+    :param str side: "source" or "target"
+    :return: word indices, int32, shape [num_words]
+    :rtype: numpy.ndarray
+    """
+    line = line.strip() + POSTFIX
+    words = line.split()
+
+    indices = [self._vocabularies[side].get(word, self._unknown_ids[side]) for word in words]
+
+    indices_numpy = numpy.array(indices, dtype=numpy.int32)
+    return indices_numpy
+
+  @staticmethod
+  def _open_file(file_name):
+    """
+    :param str file_name: filename of a plain text file, possibly zipped
+    :return: file handle
```

Comment:
Btw, this `:return:` info is kind of redundant, or at least doesn't add any more information over the `:rtype:`, which says the same, but more explicitly. In such cases, I would just leave it away...